presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -328,8 +328,8 @@ public QueryInfo createQuery(Session session, String query)
         QueryInfo queryInfo = queryExecution.getQueryInfo();
         queryMonitor.queryCreatedEvent(queryInfo);
 
-        queryExecution.addStateChangeListener(newValue -> {
-            if (newValue.isDone()) {
+        queryExecution.addQueryInfoStateChangeListener(newValue -> {
+            if (newValue.isFinalQueryInfo()) {
```

Comment:
If you document this a only firing once with the final query info, this check is not needed.